test/utils/test_knowledge_utils.py

```diff
@@ -0,0 +1,357 @@
+        assert (
+            False
+        ), "apply_resolution_strategy didn't warn on non-implemented resolution strategy"
+
+    # Test 'first' strategy
+    entity_list = ["First", "Second"]
+    entity = apply_resolution_strategy([], entity_list, "first")
+    assert entity == "First"
+
+    entity_list = [["list"], "second"]
+    entity = apply_resolution_strategy([], entity_list, "first")
+    assert entity == ["list"]
+
+    # Test 'random' strategy
+    entity_list = ["First", "Second", "Third"]
+    for _ in range(0, 3):
+        # try multiple times
+        entity = apply_resolution_strategy([], entity_list, "first")
+        assert entity in entity_list
+
+    entity_list = ["Only"]
+    entity = apply_resolution_strategy([], entity_list, "first")
+    assert entity == "Only"
+
+    # Test 'nearest' strategy
+    # Test that no robot warns
+    entity_list = ["Only"]
+    try:
+        entity = apply_resolution_strategy([], entity_list, "nearest", None)
+    except UserWarning:
+        # test passed, this should warn
+        pass
+    else:
+        assert (
+            False
+        ), "apply_resolution_strategy didn't warn on non-existent robot strategy"
+
+    robot = Robot("test_robot")
+
+    class Entity:
+        pose = Pose()
+
+    entity_list = [Entity(), Entity(), Entity()]
+    entity_list[0].pose = Pose.from_list([1, 0])
+    entity_list[1].pose = Pose.from_list([2, 0])
+    entity_list[2].pose = Pose.from_list([3, 0])
+    entity = apply_resolution_strategy([], entity_list, "nearest", robot)
+    assert entity == entity_list[0]
+
+    robot.pose = Pose.from_list([2, 1])
+    entity = apply_resolution_strategy([], entity_list, "nearest", robot)
+    assert entity == entity_list[1]
+
+    robot.pose = Pose.from_list([1000, -1000])
+    entity = apply_resolution_strategy([], entity_list, "nearest", robot)
+    assert entity == entity_list[2]
+
+
+def test_query_to_entity():
+    test_world = load_world()
+    query = "kitchen table apple"
+    entity = query_to_entity(test_world, query.split(), "location")
+    assert entity.name == "table0_tabletop"
+    entity = query_to_entity(test_world, query.split(), "object")
+    assert entity.name == "gala"
+
+    query = "kitchen table"
+    entity = query_to_entity(test_world, query.split(), "location")
+    assert entity.name == "table0"
+
+    # if we query to a random object on the kitchen table, its parent category should be table
+    entity = query_to_entity(test_world, query.split(), "object", "random")
+    assert entity.parent.category == "table"
+
+    # nearest object on the kitchen table should be the banana
+    robot = Robot("test_robot")
+    robot.pose = Pose.from_list([1.0, -0.5])
+    query = "kitchen table"
+    entity = query_to_entity(test_world, query.split(), "object", "nearest", robot)
+    assert (
+        entity.category == "banana"
+        and entity.parent.category == "table"
+        and entity.parent.parent.parent.name == "kitchen"
+    )
```

Comment:
I was originally going to suggest splitting off the asserts into separate lines, but then realized that this means if the first condition failed, the subsequent ones wouldn't even be evaluated!

I guess `pytest` doesn't have a native implementation of "expect" vs. "assert", but I found this https://pypi.org/project/pytest-assume/ ... probably not worth doing in this PR, but I should investigate later.

Reply:
Actually, I forgot I refactored things so you may only need to add it here: https://github.com/sea-bass/pyrobosim/blob/main/test/python_test_requirements.txt

Reply:
So I added it to `python_test_requirements.txt` and couldn't seem to get it to work. It got installed and I re-sourced the environment but I couldn't get it to work. `from pytest import assume` never worked, nor did `import pytest` then `pytest.assume(True)` or similar. Do I need to use `@pytest.mark.parametrize` like they do in their [examples](https://github.com/astraw38/pytest-assume)?

Reply:
I just tried locally and it worked great for me...

You can either just do `pip install pytest-assume` in your existing environment, or if you add it to the requirements file you'll need to rerun the entire `setup/create_python_env.bash`.

... but like, this doesn't matter for this PR -- it was just an out there suggestion. I'll let you follow up in another if you care!